rollbar-java/build.gradle

```diff
@@ -12,12 +12,66 @@ buildscript {
 apply plugin: "nebula.integtest"
 
 dependencies {
-    api project(':rollbar-api')
+  api project(':rollbar-api')
 
-    api group: 'org.slf4j', name: 'slf4j-api', version: '1.7.25'
+  api group: 'org.slf4j', name: 'slf4j-api', version: '1.7.25'
 
-    compileOnly 'com.google.code.findbugs:jsr305:3.0.2'
+  compileOnly 'com.google.code.findbugs:jsr305:3.0.2'
 
-    integTestImplementation group: 'com.github.tomakehurst', name: 'wiremock', version: '2.27.0'
-    integTestImplementation group: 'com.google.code.gson', name: 'gson', version: '2.8.2'
+  integTestImplementation group: 'com.github.tomakehurst', name: 'wiremock', version: '2.27.0'
+  integTestImplementation group: 'com.google.code.gson', name: 'gson', version: '2.8.2'
+}
+
+/**
+ * This task will create a version property that is statically referenced when populating the
+ * `notifier` section of the payload. It helps when users shade and / or relocate the
+ * `rollbar-java` classes, since in those cases we no longer have access to our jar manifest.
+ * The task creates a Java class instead of a text resource, since dynamically loaded resources
+ * are not as reliable under relocation as a strongly typed bytecode reference to a compiled class.
+ */
+task createVersionClass() {
+  ext.set("OUTPUT_DIR", [buildDir.getAbsolutePath(), 'src', 'generated', 'main'].join(File.separator))
+
+  outputs.dir(ext.OUTPUT_DIR)
+
+  doLast {
+    def pkg = ["com", "rollbar", "notifier", "provider", "notifier"];
+
+    def pkgName = pkg.join(".");
+    def pkgPath = "${ext.OUTPUT_DIR}${File.separator}" +
+            (['src', 'generated'] + pkg).join(File.separator)
+    def escapedVersion = VERSION_NAME.replace('"', '\\"');
+
+    def classText = """package ${pkgName};
+
+class VersionHelperResources {
+  static String getVersion() {
+    return "${escapedVersion}";
+  }
+}
+"""
+
+    new File(pkgPath).mkdirs()
+    def classFile = new File(pkgPath, 'VersionHelperResources.java').newWriter()
+    try {
+      classFile << classText
+    } finally {
+      classFile.close()
+    }
+  }
+}
+
+sourceSets {
+  main {
+    java.srcDirs += project.tasks.createVersionClass.ext.OUTPUT_DIR
+  }
+}
+
+project.tasks.compileJava.dependsOn(project.tasks.createVersionClass);
+project.tasks.checkstyleMain.dependsOn(project.tasks.createVersionClass);
+
+test {
```

Comment:
Is this needed if we are getting the version from the generated class?

Reply:
There is a test in our suite (versionReturnedShouldMatchManifestVersion below) that is basically making sure our hand-rolled code generation works, this is a way to pass the version to the test that doesn't depend on our code generation. A plain env var is a lot more reliable than our own code generation, so it is good enough as an oracle for the test.